[BLOG] Add CuPy CZI grant year 1 report blog post (#802)
Quansight/Quansight-website: apps/labs/posts/cupy-czi-grant-year1.md

  <figcaption style={{textAlign: "center"}}>Relative speedup of CuPy’s <tt>BSpline</tt>, <tt>PPoly</tt> and <tt>BPoly</tt> interpolation classes relative to SciPy for various input sizes.</figcaption>
</figure>

For those APIs which could not be readily ported from CPU to GPU, new implementations were derived from scientific papers, including the new IIR linear filtering routines, which represent the first time IIR filters can be applied on GPUs from Python. As part of this process, some enhancements and simplifications were also backported to SciPy, such as a major refactor of symmetric IIR filters to leverage IIR linear filtering and second order sections rather than dedicated routines that reduplicate code.

The following example demonstrates the new capabilities of CuPy in filtering multiple signals in parallel using a GPU. For illustrative purposes, we begin by defining four different signals.

```python
import cupy as cp
from cupyx.scipy import signal
import matplotlib.pyplot as plt

cp.random.seed(2354)

> rng = cp.random.default_rng(seed=1234)

Reviewer: Is it necessary to set two different seeds here?

Author: If I used the same seed in both the global and the particular generators, the signals produced don't have much variety, hence the two seeds

t = cp.linspace(-1, 1, 201)
t = cp.broadcast_to(t, (4, 201)).copy()
phase = 10 * cp.random.rand(4, 1) - 5

# Define 4 random signals composed of a main component with quadratic time
# input and two sinusoids with 1.25Hz and 3.85Hz frequencies, plus a DC
# component.
main_x = cp.sin(2 * cp.pi * 0.75 * t * (1 - t) + 2.1 + phase)
```

<figure>
  <img src="/posts/cupy-czi-grant-year1/original_signals.png" alt="A grid of four plots, arranged in a 2x2 configuration. Each plot denotes a signal in the time domain, the signals where generated by a sine wave with quadratic time input and two sinusoids with 1.25Hz and 3.85Hz frequencies, plus a DC component" />
  <figcaption style={{textAlign: "center"}}>Four random signals composed of a main component with quadratic time
input and two sinusoids with 1.25Hz and 3.85Hz frequencies, plus a DC
component.</figcaption>

> </figure>

Reviewer 2: This takes up a full page on my 32" monitor, as do the next figures. It would look better if it were 1.5x or 2x smaller - could you try and shrink the size a bit?

Reviewer: Shouldn't the figure sizes be adjusted via CSS, not by literally making them smaller? Typically you want images as high-resolution as possible, otherwise you get aliasing on high resolution monitors.

Reviewer 2: I meant "visually smaller". And sure, using CSS and reasonable file sizes for PNG/JPG images makes sense.

# Apply the filter again.
z2, _ = signal.lfilter(b, a, z, zi=zi*z[:, 0, None])

# Use filtfilt on all signals in parallel.
y = signal.filtfilt(b, a, xn)
```

A plot comparing filtering results is displayed below. By visual inspection (and rather cursory analysis!), we can conclude that `filtfilt` does a better job at recovering the original signals.

<figure>
  <img src="/posts/cupy-czi-grant-year1/filtered_signals.png" alt="A grid of four plots, arranged in a 2x2 configuration. Each plot shows the result of taking a noisy signal, applying the lfilter and filtfilt CuPy in the following order: lfilter once, lfilter twice and filfilt. The filtfilt results reconstruct the original signal with higher fidelity." />
  <figcaption style={{textAlign: "center"}}>Filtering results using the new lfilter and filtfilt CuPy APIs. Each result is compared against the original, non-noisy signal.</figcaption>
</figure>

As part of this effort, we made several other improvements to other CuPy modules, such as adding special functions for elliptic filters, implementing the matrix exponential in the `cupyx.scipy.linalg` namespace, and implementing KD-Trees in `cupyx.scipy.spatial`.

> This work would not be possible without the help of the [cuSignal](https://github.com/rapidsai/cusignal) team at the NVIDIA RAPIDS initiative, where more than 36 APIs will be migrated from their project into CuPy, thus unifying the GPU signal processing efforts currently available in the ecosystem. As part of this migration, cuSignal will be deprecated and replaced by the new `cupyx.scipy.signal` module calls. The calls will be interchangeable by simply replacing the `cusignal` imports with imports from `cupyx.scipy.signal`. For more information about this collaboration, we invite you to visit the [RAPIDS blog](https://medium.com/rapids-ai).

Reviewer 2: Can you link to the actual blog post from the RAPIDS team?

Author: I think they haven't published it yet, that's why I left the link to their overall blog instead to the particular page, since it does not exist yet. Let me confirm with @awthomp if the blog post is now available

Reviewer 2: It isn't there yet. @awthomp do you know if is it expected to arrive shortly? If so, we can wait for it - or we can just leave out this sentence, and add it back in if and when that post appears. Right now the reader is invited to go look for something that isn't there, which would be good to avoid.

Author: I'll remove the sentence for the sake of publishing

As the grant enters its second year, we plan to finish the overall interoperability of the interpolation module, which first requires a significant refactor of several FORTRAN routines for spline smoothing into GPU kernels and, second, implementing computational geometry routines to compute convex hulls and Delaunay triangulations, which are used for interpolating multi-dimensional unstructured data. This work will further enhance overall CuPy-SciPy compatibility.

These new APIs will open a sea of possibilities in terms of acceleration and batch processing of data, and we hope they will be useful to the CuPy community. Right now, these additions are being pre-released as part of the beta releases of the upcoming CuPy v13.0. We invite you to keep an eye on our tracking issues https://github.com/cupy/cupy/issues/7186 and https://github.com/cupy/cupy/issues/7403, where you can find more information regarding the implementation discussions, as well as the current progress on the remaining functionality of CuPy’s interpolation and signal modules. Please feel free to chime in and contribute to any discussion or improvement to the project!

> We would like to thank the [Chan Zuckerberg Initiative (CZI)](https://chanzuckerberg.com/eoss/), Preferred Networks, and Quansight Labs for the resources used to plan and execute the current grant. We would also like to thank NVIDIA and the RAPIDS team for their help and willingness to migrate cuSignal into CuPy as part of this work.

Reviewer 2: This looks slightly off with CZI being a link and Preferred Networks and following entities not. I'd suggest making them all links, or none.

Also, this is the Labs blog, so no need to thank ourselves here I'd say? The mention in the next paragraph is probably enough.

---
title: 'Improving the interpolation and signal processing capabilities of CuPy'

> published: November 2, 2023

Reviewer 2: This date should be updated right before merging the post.

y = signal.filtfilt(b, a, xn)
```

A plot comparing filtering results is displayed below. By visual inspection
(and rather cursory analysis!), we can conclude that `filtfilt` does a better
job at recovering the original signals.

<figure>
  <img
    src="/posts/cupy-czi-grant-year1/filtered_signals.png"
    alt="A grid of four plots, arranged in a 2x2 configuration. Each plot shows
    the result of taking a noisy signal, applying the lfilter and filtfilt CuPy
    in the following order: lfilter once, lfilter twice and filfilt. The filtfilt
    results reconstruct the original signal with higher fidelity."
    style={{maxWidth: "70%", display: "block", marginLeft: "auto", marginRight: "auto"}}
  />
  <figcaption style={{textAlign: "center"}}>
    Filtering results using the new lfilter and filtfilt CuPy APIs.
    Each result is compared against the original, non-noisy signal.
  </figcaption>
</figure>

As part of this effort, we made several other improvements to other CuPy modules,
such as adding special functions for elliptic filters, implementing the
matrix exponential in the `cupyx.scipy.linalg` namespace, and implementing
KD-Trees in `cupyx.scipy.spatial`.

> This work would not be possible without the help of the

Reviewer 2: ```suggestion
This work would not have been possible without the help of the
```

[cuSignal](https://github.com/rapidsai/cusignal) team at the NVIDIA RAPIDS

> initiative, where more than 36 APIs will be migrated from their project into

Reviewer 2: ```suggestion
initiative, who agreed to migrate the implementations of over 36 APIs from their project to
```

replaced by the new `cupyx.scipy.signal` module calls. The calls will be
interchangeable by simply replacing the `cusignal` imports with imports from
`cupyx.scipy.signal`. For more information about this collaboration,
we invite you to visit the [RAPIDS blog](https://medium.com/rapids-ai).

As the grant enters its second year, we plan to finish the overall
interoperability of the interpolation module, which first requires a
significant refactor of several FORTRAN routines for spline smoothing into GPU
kernels and, second, implementing computational geometry routines to compute
convex hulls and Delaunay triangulations, which are used for interpolating
multi-dimensional unstructured data. This work will further enhance overall
CuPy-SciPy compatibility.

These new APIs will open a sea of possibilities in terms of acceleration and
batch processing of data, and we hope they will be useful to the CuPy community.
Right now, these additions are being pre-released as part of the beta releases
of the upcoming CuPy v13.0. We invite you to keep an eye on our tracking issues
https://github.com/cupy/cupy/issues/7186 and https://github.com/cupy/cupy/issues/7403,

> where you can find more information regarding the implementation discussions,

Reviewer 2: ```suggestion
where you can find more information on and discussions about implementations,
```
